nav2_bt_navigator/src/bt_navigator.cpp

```diff
@@ -206,22 +210,59 @@ void
 BtNavigator::onLoop()
 {
   // action server feedback (pose, duration of task,
-  // number of recoveries, and distance remaining to goal)
+  // time remaining, number of recoveries, and
+  // distance remaining to goal)
   auto feedback_msg = std::make_shared<Action::Feedback>();
 
+  geometry_msgs::msg::PoseStamped current_pose;
   nav2_util::getCurrentPose(
-    feedback_msg->current_pose, *tf_, global_frame_, robot_frame_, transform_tolerance_);
+    current_pose, *tf_, global_frame_, robot_frame_, transform_tolerance_);
 
   auto blackboard = bt_action_server_->getBlackboard();
 
-  geometry_msgs::msg::PoseStamped goal_pose;
-  blackboard->get(goal_blackboard_id_, goal_pose);
-
-  feedback_msg->distance_remaining = nav2_util::geometry_utils::euclidean_distance(
-    feedback_msg->current_pose.pose, goal_pose.pose);
+  // Get current path points
+  nav_msgs::msg::Path current_path;
+  blackboard->get("path", current_path);
+
+  // Find the closest pose to current pose on global path
+  auto find_closest_pose_idx =
+    [&current_pose, &current_path]() {
+      size_t closest_pose_idx = 0;
+      double curr_min_dist = std::numeric_limits<double>::max();
+      for (size_t curr_idx = 0; curr_idx < current_path.poses.size(); ++curr_idx) {
+        double curr_dist = nav2_util::geometry_utils::euclidean_distance(
+          current_pose, current_path.poses[curr_idx]);
+        if (curr_dist < curr_min_dist) {
+          curr_min_dist = curr_dist;
+          closest_pose_idx = curr_idx;
+        }
+      }
+      return closest_pose_idx;
+    };
+
+  // Calculate distance on the path
+  double distance_remaining =
+    nav2_util::geometry_utils::calculate_path_length(current_path, find_closest_pose_idx());
+
+  // Default value for time remaining
+  rclcpp::Duration estimated_time_remaining = rclcpp::Duration::from_seconds(0.0);
+
+  // Get current speed
+  geometry_msgs::msg::Twist current_odom = odom_smoother_->getTwist();
+  double current_linear_speed = hypot(current_odom.linear.x, current_odom.linear.y);
```

Comment:
Use `std::hypot` with the namespace for clarity

Reply:
Agreed! thanks for the suggestion.